File: airesources/Swift/Sources/hlt/Navigation.swift
import Foundation

/// Creates a move that navigates a ship to a given dock target
public func navigateShipToDock(map: Map, ship: Ship, dockTarget: Entity, maxThrust: Int) -> Move {
    let angularStepRad = Double.pi / 180.0
    let targetPosition = ship.getClosestPoint(to: dockTarget)
    
    return navigateShipTowardsTarget(map: map,
                                     ship: ship,
                                     target: targetPosition,
                                     maxThrust: maxThrust,
                                     avoidObstacles: true,
                                     maxCorrections: Constants.MaxNavigationCorrections,
                                     angularStepRad: angularStepRad)
}

/// Creates a move that navigates a ship to a given target
public func navigateShipTowardsTarget(map: Map,
                                      ship: Ship,
                                      target: Entity,
                                      maxThrust: Int,
                                      avoidObstacles: Bool,
                                      maxCorrections: Int,
                                      angularStepRad: Double) -> Move {
    if maxCorrections <= 0 {
        return .noop
    }
    
    let distance = ship.distance(to: target)
    let angleRad = ship.orientTowardsInRadians(x: target.x, y: target.y)
    
    if avoidObstacles && map.obstaclesBetween(start: ship, target: target).count > 0 {
        let newTargetDx = cos(angleRad + angularStepRad) * distance
        let newTargetDy = sin(angleRad + angularStepRad) * distance
        let newTarget = Position(x: ship.x + newTargetDx, y: ship.y + newTargetDy)
        
        return navigateShipTowardsTarget(map: map,
                                         ship: ship,
                                         target: newTarget,
                                         maxThrust: maxThrust,
                                         avoidObstacles: avoidObstacles,
                                         maxCorrections: maxCorrections - 1,
                                         angularStepRad: angularStepRad)
    }
    
    let thrust: Int
    
    if distance < Double(maxThrust) {
        thrust = Int(distance)
    } else {
        thrust = maxThrust
    }
    
    let angleDeg = angleRad.radiansToDegrees.truncatingRemainder(dividingBy: 360)
    
    return .thrust(ship, angleDeg: Int(angleDeg), thrust: thrust)
}

File: airesources/Dart/lib/src/navigation.dart
part of hlt;


ThrustMove navigateShipToDock(
    GameMap gameMap,
    Ship ship,
    Entity dockTarget,
    int maxThrust) {
  final maxCorrections = MAX_NAVIGATION_CORRECTIONS;
  final avoidObstacles = true;
  final angularStepRad = PI/180.0;
  final targetPos = ship.getClosestPoint(dockTarget);
  return navigateShipTowardsTarget(gameMap, ship, targetPos, maxThrust, avoidObstacles, maxCorrections, angularStepRad);
}

ThrustMove navigateShipTowardsTarget(
    GameMap gameMap,
    Ship ship,
    Position targetPos,
    int maxThrust,
    bool avoidObstacles,
    int maxCorrections,
    double angularStepRad) {
  if (maxCorrections <= 0) return null;
  final distance = ship.getDistanceTo(targetPos);
  final angleRad = ship.orientTowardsInRad(targetPos);
  if (avoidObstacles && gameMap.objectsBetween(ship, targetPos).isNotEmpty) {
    final newTargetDx = cos(angleRad + angularStepRad) * distance;
    final newTargetDy = sin(angleRad + angularStepRad) * distance;
    final newTarget = new Position(x: ship.x + newTargetDx, y: ship.y + newTargetDy);
    return navigateShipTowardsTarget(gameMap, ship, newTarget, maxThrust, true, (maxCorrections-1), angularStepRad);
  }
  final int thrust = (distance < maxThrust ? distance : maxThrust).toInt();
  final int angleDeg = angleRadToDegClipped(angleRad);
  return new ThrustMove(ship.id, thrust, angleDeg);
}

bool segmentCircleIntersect(Position start, Position end, Entity circle, double fudge) {
  // Parameterize the segment as start + t * (end - start),
  // and substitute into the equation of a circle
  // Solve for t
  final double circleRadius = circle.radius;
  final double startX = start.x;
  final double startY = start.y;
  final double endX = end.x;
  final double endY = end.y;
  final double centerX = circle.x;
  final double centerY = circle.y;
  final double dx = endX - startX;
  final double dy = endY - startY;

  final double a = pow(dx, 2) + pow(dy, 2);

  final double b = -2 * (pow(startX, 2) - (startX * endX)
    - (startX * centerX) + (endX * centerX)
    + pow(startY, 2) - (startY * endY)
    - (startY * centerY) + (endY * centerY));

  if (a == 0.0) {
    // Start and end are the same point
    return start.getDistanceTo(circle) <= circleRadius + fudge;
  }

  // Time along segment when closest to the circle (vertex of the quadratic)
  final double t = min(-b / (2 * a), 1.0);
  if (t < 0) return false;

  final double closestX = startX + dx * t;
  final double closestY = startY + dy * t;
  final double closestDistance = new Position(x: closestX, y: closestY).getDistanceTo(circle);
  return closestDistance <= circleRadius + fudge;
}
